File: packages/nodejs/codepotx/render-clients/handlebars/stdio.mjs
import { createHash } from 'node:crypto';
import readline from 'node:readline';
import Handlebars from 'handlebars';

const RENDERER_ID = 'handlebars';
const RENDERER_VERSION = '4.7.9';
const CAPABILITY = 'handlebars/v1';
const revision = process.env.DRYV_RENDERER_REVISION ?? 'default';
const fingerprint = sha256(`dryv-handlebars\0${RENDERER_VERSION}\0${revision}\0strict-v1`);
const cancelled = new Set();

const input = readline.createInterface({ input: process.stdin, crlfDelay: Infinity });

for await (const line of input) {
  let response;
  try {
    const message = JSON.parse(line);
    response = handle(message);
  } catch (error) {
    response = {
      type: 'protocol-error',
      code: 'HANDLEBARS_PROTOCOL_ERROR',
      message: error instanceof Error ? error.message : String(error),
    };
  }
  process.stdout.write(`${JSON.stringify(response)}\n`);
  if (response.type === 'shutdown-result') process.exit(0);
}

function handle(message) {
  if (!message || typeof message !== 'object' || Array.isArray(message)) {
    throw new Error('protocol message must be an object');
  }
  if (message.type === 'hello') {
    return {
      type: 'hello-result',
      rendererId: RENDERER_ID,
      rendererVersion: RENDERER_VERSION,
      capabilities: [CAPABILITY],
      protocolVersions: [1],
      contextVersions: [1],
      templateMediaTypes: ['text/x-handlebars-template', 'application/x-handlebars-template'],
      fingerprint,
      maxConcurrency: 1,
      deterministic: true,
    };
  }
  if (message.type === 'cancel') {
    requireString(message.jobId, 'jobId');
    cancelled.add(message.jobId);
    return { type: 'cancel-result', jobId: message.jobId };
  }
  if (message.type === 'shutdown') return { type: 'shutdown-result' };
  if (message.type !== 'render') throw new Error(`unknown protocol message type ${String(message.type)}`);

  const request = message.request;
  if (!request || typeof request !== 'object' || Array.isArray(request)) throw new Error('render request must be an object');
  const jobId = requireString(request.jobId, 'jobId');
  if (cancelled.has(jobId)) {
    return { type: 'render-result', jobId, rendererFingerprint: fingerprint, outputs: [], diagnostics: [], cancelled: true };
  }
  if (!Array.isArray(request.outputs) || request.outputs.length !== 1) {
    return {
      type: 'render-result',
      jobId,
      rendererFingerprint: fingerprint,
      outputs: [],
      diagnostics: [{ code: 'HANDLEBARS_OUTPUT_CARDINALITY', message: 'Handlebars Render Client requires one planned output', severity: 'error' }],
      cancelled: false,
    };
  }

  try {
    const source = Buffer.from(requireString(request.templateContentBase64, 'templateContentBase64'), 'base64').toString('utf8');
    const template = Handlebars.compile(source, { strict: true, noEscape: true });
    const rendered = Buffer.from(template(request.context ?? {}), 'utf8');
    const outputId = requireString(request.outputs[0]?.id, 'output.id');
    return {
      type: 'render-result',
      jobId,
      rendererFingerprint: fingerprint,
      outputs: [{
        id: outputId,
        contentBase64: rendered.toString('base64'),
        contentHash: artifactHash(rendered),
      }],
      diagnostics: [],
      cancelled: false,
    };
  } catch (error) {
    return {
      type: 'render-result',
      jobId,
      rendererFingerprint: fingerprint,
      outputs: [],
      diagnostics: [{
        code: 'HANDLEBARS_RENDER_FAILED',
        message: error instanceof Error ? error.message : String(error),
        severity: 'error',
      }],
      cancelled: false,
    };
  }
}

function requireString(value, label) {
  if (typeof value !== 'string' || value.length === 0) throw new Error(`${label} must be a non-empty string`);
  return value;
}

function sha256(value) {
  return `sha256:${createHash('sha256').update(value).digest('hex')}`;
}

function artifactHash(content) {
  return `sha256:v1:artifact-content:${createHash('sha256').update(content).digest('hex')}`;
}

File: packages/python/.helpers/dryv-template-jinja/src/dryv_template_jinja/stdio.py
from __future__ import annotations

import base64
import json
import sys
from dataclasses import dataclass
from typing import Mapping

from .session import JinjaRenderSession


@dataclass(frozen=True, slots=True)
class _Output:
    id: str
    path: str


@dataclass(frozen=True, slots=True)
class _Request:
    job_id: str
    template_content: bytes
    context: Mapping[str, object]
    outputs: tuple[_Output, ...]


def main() -> int:
    session = JinjaRenderSession(max_concurrency=1)
    for line in sys.stdin:
        try:
            message = json.loads(line)
            response = _handle(session, message)
        except Exception as exc:  # noqa: BLE001 - process boundary must stay alive
            response = {
                "type": "protocol-error",
                "code": "JINJA_PROTOCOL_ERROR",
                "message": str(exc) or type(exc).__name__,
            }
        sys.stdout.write(json.dumps(response, separators=(",", ":"), sort_keys=True) + "\n")
        sys.stdout.flush()
        if response.get("type") == "shutdown-result":
            return 0
    return 0


def _handle(session: JinjaRenderSession, message: object) -> dict[str, object]:
    if not isinstance(message, dict):
        raise ValueError("protocol message must be an object")
    kind = message.get("type")
    if kind == "hello":
        hello = session.hello()
        return {
            "type": "hello-result",
            "rendererId": hello.renderer_id,
            "rendererVersion": hello.renderer_version,
            "capabilities": list(hello.capabilities),
            "protocolVersions": list(hello.protocol_versions),
            "contextVersions": list(hello.context_versions),
            "templateMediaTypes": list(hello.template_media_types),
            "fingerprint": hello.fingerprint,
            "maxConcurrency": hello.max_concurrency,
            "deterministic": hello.deterministic,
        }
    if kind == "cancel":
        job_id = _string(message.get("jobId"), "jobId")
        session.cancel(job_id)
        return {"type": "cancel-result", "jobId": job_id}
    if kind == "shutdown":
        return {"type": "shutdown-result"}
    if kind != "render":
        raise ValueError(f"unknown protocol message type {kind!r}")

    raw = message.get("request")
    if not isinstance(raw, dict):
        raise ValueError("render request must be an object")
    try:
        template = base64.b64decode(_string(raw.get("templateContentBase64"), "templateContentBase64"), validate=True)
    except Exception as exc:  # noqa: BLE001
        raise ValueError("templateContentBase64 is invalid") from exc
    context = raw.get("context")
    if not isinstance(context, dict) or not all(isinstance(key, str) for key in context):
        raise ValueError("context must be a string-keyed object")
    outputs_raw = raw.get("outputs")
    if not isinstance(outputs_raw, list):
        raise ValueError("outputs must be an array")
    outputs = tuple(
        _Output(_string(item.get("id"), "output.id"), _string(item.get("path"), "output.path"))
        for item in outputs_raw
        if isinstance(item, dict)
    )
    if len(outputs) != len(outputs_raw):
        raise ValueError("outputs must contain objects")
    request = _Request(
        _string(raw.get("jobId"), "jobId"),
        template,
        context,
        outputs,
    )
    result = session.render(request)
    return {
        "type": "render-result",
        "jobId": result.job_id,
        "rendererFingerprint": result.renderer_fingerprint,
        "outputs": [
            {
                "id": item.id,
                "contentBase64": base64.b64encode(item.content).decode("ascii"),
                "contentHash": item.content_hash,
            }
            for item in result.outputs
        ],
        "diagnostics": [
            {"code": item.code, "message": item.message, "severity": item.severity}
            for item in result.diagnostics
        ],
        "cancelled": result.cancelled,
    }


def _string(value: object, label: str) -> str:
    if not isinstance(value, str) or not value:
        raise ValueError(f"{label} must be a non-empty string")
    return value


if __name__ == "__main__":
    raise SystemExit(main())
